Launch security with AMD SEV-SNP

ymjing opened the pull request:
### What this PR does
Before this PR:
No SEV-SNP guests are supported. 

After this PR:
The node controller 
SEV-SNP guests can run with `launchSecutity --> sev --> policy --> secureNestedPaging: true`.

### Why we need it and why it was done in this way
SEV-SNP is different from SEV or SEV-ES in many ways, such as the guest policy bits, libvirt launchSecurity format. Thus I'd suggest we add a dedicated `secureNestedPaging` type.

The following tradeoffs were made:

The following alternatives were considered:

### Special notes for your reviewer
The PR makes the following changes to the existing kubevirt componennts:

1. `virt-handler`: the node labeller now supports SEV-SNP capable nodes;
2. `virt-launcher`:  the converter now generates SNP-specific domxml config for libvirt.

This PR works on top of `kubevirt 0.15.0-beta0`, which uses `libvirt 10.10.0-3.el9`  and `qemu-kvm 9.1.0-7el9`.

However, the firmware in `edk2-ovmf 20241117-1.el9` does not seem to work. For testing I have to use the `OVMF.fd` from Debian Trixie and the hook sidecar to patch the domxml with a Python script like this.

```python
    #!/usr/bin/env python3

    import xml.etree.ElementTree as ET
    import sys

    def main(domxml):
        root = ET.fromstring(domxml)

        os = root.find("os")
        if os is not None:
            loader = os.find("loader")
            if loader is not None:
                # Replace 'pflash' with 'rom' as SNP can't use pflash loader
                loader.set("type", "rom")
                # Use a custom firmware
                loader.text = "/var/run/firmware/OVMF.fd"
            nvram = os.find("nvram")
            if nvram is not None:
                # Remove the unused OVMF_VAR.fd
                os.remove(nvram)
        ET.dump(root)

    if __name__ == "__main__":
        main(sys.argv[4])
``` 
 
### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Experimental support for AMD SEV-SNP
```

alicefr commented:
@ymjing after a first pass, we will need a feature gate and at least a functional test

ymjing commented:
> @ymjing after a first pass, we will need a feature gate and at least a functional test

Hi Alice, thanks for getting back to me. Would a new `WorkloadEncryptionSEVSNP` feature gate soound ok to you?  `WorkloadEncryptionSEVSNP` implicitly enables `WorkloadEncryptionSEV`.

ymjing commented:
The cloud images that are verified to work with SEV-SNP on this PR are:

* Fedora Cloud Base 41 (non-UKI): [download link](https://download.fedoraproject.org/pub/fedora/linux/releases/41/Cloud/x86_64/images/Fedora-Cloud-Base-Generic-41-1.4.x86_64.qcow2); (note: `autoattachGraphicsDevice` has to be `false`)

* CentOS Stream 10 (EFI): [download link](https://cloud.centos.org/centos/10-stream/x86_64/images/CentOS-Stream-GenericCloud-x86_64-10-latest.x86_64.qcow2)

alicefr commented:
> > @ymjing after a first pass, we will need a feature gate and at least a functional test
> 
> Hi Alice, thanks for getting back to me. Would a new `WorkloadEncryptionSEVSNP` feature gate sound ok to you? Say`WorkloadEncryptionSEVSNP` would implicitly enable `WorkloadEncryptionSEV`.

Yes, but please don't enable other feature gates implicitly. This needs either be documented or we could graduate the feature gate.

alicefr commented:
Can you please add more information on the attestation? Is the VM attested and if yes how

ymjing commented:
> Can you please add more information on the attestation? Is the VM attested and if yes how

No, this PR does not include SEV-SNP attestation. The community seems to already have some thoughts on how ES / SNP attestation should be implemented (https://github.com/kubevirt/kubevirt/issues/6991). I'd prefer to land attestation as a separate PR. Wdyt?

ymjing commented:
> > > @ymjing after a first pass, we will need a feature gate and at least a functional test
> > 
> > 
> > Hi Alice, thanks for getting back to me. Would a new `WorkloadEncryptionSEVSNP` feature gate sound ok to you? Say`WorkloadEncryptionSEVSNP` would implicitly enable `WorkloadEncryptionSEV`.
> 
> Yes, but please don't enable other feature gates implicitly. This needs either be documented or we could graduate the feature gate.

Can I still put SEV-SNP feature behind the existing `WorkloadEncryptionSEV` feature gate? I tried to add a separate` WorkloadEncryptionSEVSNP` feature gate today; and soon I found myself basically polluting the codebase with `WorkloadEncryptionSEVSNP` whereever `WorkloadEncryptionSEV` is used. 

And it seems a bit counter-intuitive for users becuase (1) SEV-ES does not have its own feature gate; and (2) `secureNestedPaging` is at the same level as `encryptedState` under `sev` --> like this 

```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-centos-snp
  name: vmi-centos-snp
spec:
  domain:
    launchSecurity:
      sev:
        policy:
          secureNestedPaging: true
    firmware:
      bootloader:
        efi:
          secureBoot: false
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
      inputs: []
      autoattachMemBalloon: false
      autoattachGraphicsDevice: false
      autoattachSerialConsole: true
      autoattachPodInterface: true
      autoattachVSOCK: true
      disableHotplug: true
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: bincat01/centos-stream-10:snp
    name: containerdisk
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        chpasswd:
          list: |
            root:centos
          expire: False
    name: cloudinitdisk
```

ymjing commented:
Re: func tests, I'm blocked becuase the current `fedora-with-test-tooling` does not use a guest kernel that supports SEV-SNP guests. Two options: (1) upgrade the base image of `fedora-with-test-tooling` to Fedora 41; (2) introduce a separate containder disk for testing SEV / SEV-ES / SEV-SNP features. Personally I prefer (2). And I'd like to know what the kubevirt community suggests. Thanks for your time reading this.

zhencliu commented:
hi @ymjing , thank you for the patch. May I ask the timeline of the patch? And what's the priority? 
BTW: How can I try it in my local environment? Thanks.

ymjing commented:
> hi @ymjing , thank you for the patch. May I ask the timeline of the patch? And what's the priority? BTW: How can I try it in my local environment? Thanks.

Appreciate it if I can land this PR before the end of Q1. As we hope to run confidential VMs in production on OpenShift w/ Virtualization in Q2.

To test it you'd need a server/workstation that supports AMD SEV-SNP. The CPU has to be AMD EPYC Milan (7xx3) or later generations; and the motherboard / BIOS needs to support SEV-SNP, too. For example, Supermicro 3014TS-i + AMD EPYC 7xx3; and Dell R6615 + EPYC 9xx4.

ymjing commented:
> Hi @ymjing , I hit the error when trying to create a snp vm by 'kubectl apply -f snp.yaml" command: snippet of snp.yaml `spec: domain: launchSecurity: sev: policy: secureNestedPaging: true`
> 
> Error from server (BadRequest): error when creating "snp.yaml": VirtualMachine in version "v1" cannot be handled as a VirtualMachine: strict decoding error: unknown field "spec.template.spec.domain.launchSecurity.sev.policy.secureNestedPaging"
> 
> It seems the secureNestedPaging is not recognized, does it need to be added in somewhere, e.g. api/openapi-spec/swagger.json, just like encryptedState? Thanks.

I've added `secureNestedPaging` to the API in [staging/src/kubevirt.io/api/core/v1/schema.go](https://github.com/kubevirt/kubevirt/pull/13755/commits/db2a4d439bce63e53e864aebbd2fae9969cbde72#diff-df03d2826ecf8b2a2c21bbfec2ec08685fbb7ce0486696ccbfdd074f5e50c82c); I recall that an extra step is needed to re-generated the updated Kubevirt APIs -  might be `make generate`?

zhencliu commented:
> > Hi @ymjing , I hit the error when trying to create a snp vm by 'kubectl apply -f snp.yaml" command: snippet of snp.yaml `spec: domain: launchSecurity: sev: policy: secureNestedPaging: true`
> > Error from server (BadRequest): error when creating "snp.yaml": VirtualMachine in version "v1" cannot be handled as a VirtualMachine: strict decoding error: unknown field "spec.template.spec.domain.launchSecurity.sev.policy.secureNestedPaging"
> > It seems the secureNestedPaging is not recognized, does it need to be added in somewhere, e.g. api/openapi-spec/swagger.json, just like encryptedState? Thanks.
> 
> I've added `secureNestedPaging` to the API in [staging/src/kubevirt.io/api/core/v1/schema.go](https://github.com/kubevirt/kubevirt/pull/13755/commits/db2a4d439bce63e53e864aebbd2fae9969cbde72#diff-df03d2826ecf8b2a2c21bbfec2ec08685fbb7ce0486696ccbfdd074f5e50c82c); I recall that an extra step is needed to re-generated the updated Kubevirt APIs - might be `make generate`?

Thanks, yes, it was automatically generated after `make generate`.
Another thing is the loader setting in libvirt xml, could it use the stateless firmware OVMF.amdsev.fd instead of OVMF_CODE.cc.fd, OVMF.amdsev.fd is present under /usr/share/edk2/ovmf, which is used for amd sev. Then sidecar is not needed to run a snp vm.

ymjing commented:
> > > > Hi @ymjing , I hit the error when trying to create a snp vm by 'kubectl apply -f snp.yaml" command: snippet of snp.yaml `spec: domain: launchSecurity: sev: policy: secureNestedPaging: true`
> > > > Error from server (BadRequest): error when creating "snp.yaml": VirtualMachine in version "v1" cannot be handled as a VirtualMachine: strict decoding error: unknown field "spec.template.spec.domain.launchSecurity.sev.policy.secureNestedPaging"
> > > > It seems the secureNestedPaging is not recognized, does it need to be added in somewhere, e.g. api/openapi-spec/swagger.json, just like encryptedState? Thanks.
> > > 
> > > 
> > > I've added `secureNestedPaging` to the API in [staging/src/kubevirt.io/api/core/v1/schema.go](https://github.com/kubevirt/kubevirt/pull/13755/commits/db2a4d439bce63e53e864aebbd2fae9969cbde72#diff-df03d2826ecf8b2a2c21bbfec2ec08685fbb7ce0486696ccbfdd074f5e50c82c); I recall that an extra step is needed to re-generated the updated Kubevirt APIs - might be `make generate`?
> > 
> > 
> > Thanks, yes, it was automatically generated after `make generate`. Another thing is the loader setting in libvirt xml, could it use the stateless firmware OVMF.amdsev.fd instead of OVMF_CODE.cc.fd, OVMF.amdsev.fd is present under /usr/share/edk2/ovmf, which is used for amd sev. Then sidecar is not needed to run a snp vm.
> 
> Just FYI, I changed to use OVMF.amdsev.fd, it works.

Yeah; the `OVMF.amdsev.fd` in the `edk2-ovmf` RPM pacakge works too. I do have a seperate patch which gets rid of the sidecar; however I'm not 100% sure if SEV and SEV-ES VMs should use `OVMF.amdsev.fd` too. At least for now Kubevirt uses `OVMF_CODE.cc.fd`?

zhencliu commented:
> > > > > Hi @ymjing , I hit the error when trying to create a snp vm by 'kubectl apply -f snp.yaml" command: snippet of snp.yaml `spec: domain: launchSecurity: sev: policy: secureNestedPaging: true`
> > > > > Error from server (BadRequest): error when creating "snp.yaml": VirtualMachine in version "v1" cannot be handled as a VirtualMachine: strict decoding error: unknown field "spec.template.spec.domain.launchSecurity.sev.policy.secureNestedPaging"
> > > > > It seems the secureNestedPaging is not recognized, does it need to be added in somewhere, e.g. api/openapi-spec/swagger.json, just like encryptedState? Thanks.
> > > > 
> > > > 
> > > > I've added `secureNestedPaging` to the API in [staging/src/kubevirt.io/api/core/v1/schema.go](https://github.com/kubevirt/kubevirt/pull/13755/commits/db2a4d439bce63e53e864aebbd2fae9969cbde72#diff-df03d2826ecf8b2a2c21bbfec2ec08685fbb7ce0486696ccbfdd074f5e50c82c); I recall that an extra step is needed to re-generated the updated Kubevirt APIs - might be `make generate`?
> > > 
> > > 
> > > Thanks, yes, it was automatically generated after `make generate`. Another thing is the loader setting in libvirt xml, could it use the stateless firmware OVMF.amdsev.fd instead of OVMF_CODE.cc.fd, OVMF.amdsev.fd is present under /usr/share/edk2/ovmf, which is used for amd sev. Then sidecar is not needed to run a snp vm.
> > 
> > 
> > Just FYI, I changed to use OVMF.amdsev.fd, it works.
> 
> Yeah; the `OVMF.amdsev.fd` in the `edk2-ovmf` RPM pacakge works too. I do have a seperate patch which gets rid of the sidecar; however I'm not 100% sure if SEV and SEV-ES VMs should use `OVMF.amdsev.fd` too. At least for now Kubevirt uses `OVMF_CODE.cc.fd`?

Confirmed with my associates who are working on qemu/libvirt, for SNP only OVMF.amdsev.fd can be used, the latest configuration from xml:  
<loader type='rom'>/usr/share/edk2/ovmf/OVMF.amdsev.fd</loader>, 
but for SEV/SEV-ES, both are OK, but libvirt uses  OVMF_CODE.fd/OVMF_VARS.fd by default. 
So for SEV/SEV-ES, it might be better to keep the current logic, but for SNP, switch to OVMF.amdsev.fd 
BTW, would you please add your patch for the FW issue into this one as a seperate commit?

ymjing commented:
Patchset pushed; a SEV-SNP VMI looks like this without using a sidecar

```yaml
---
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  labels:
    special: vmi-centos-snp
  name: vmi-centos-snp
spec:
  domain:
    launchSecurity:
      sev:
        policy:
          secureNestedPaging: true
    firmware:
      bootloader:
        efi:
          secureBoot: false
    devices:
      disks:
      - disk:
          bus: virtio
        name: containerdisk
      - disk:
          bus: virtio
        name: cloudinitdisk
      rng: {}
      inputs: []
      autoattachMemBalloon: false
      autoattachGraphicsDevice: false
      autoattachSerialConsole: true
      autoattachPodInterface: true
      autoattachVSOCK: true
      disableHotplug: true
    resources:
      requests:
        memory: 1024M
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: bincat01/centos-stream-10:snp
    name: containerdisk
  - cloudInitNoCloud:
      userData: |-
        #cloud-config
        chpasswd:
          list: |
            root:centos
          expire: False
        pacakges:
          - nginx
        runcmd:
          - [ systemctl, enable, nginx ]
          - [ systemctl, start, nginx ]
    name: cloudinitdisk

```

ymjing commented:
Review requested. Glad this PR works for you @zhencliu.

The PR is missing the functional tests. See my [comment](https://github.com/kubevirt/kubevirt/pull/13755#issuecomment-2610589404) above. Kubevirt runs functional tests with the `fedora-with-test-tooling` image, which does not use an up-to-date Linux kernel that supports SEV-SNP. 

My proposal is to create a new `fedora-with-sev-tooling` image for testing everything behind the `WorkloadEncryptionSEV` feature gate, including this PR. However this requires a new PR. And I'd hope we could merge this PR first without being blocked.

> I verified this patch in my local setup, everything works fine, the patch is still in draft, is there any other blocker?

ymjing commented:
Gentle ping.

zhencliu commented:
> Gentle ping.

hi @ymjing , some suggestions about the code review from some experts, just fyi. You could send an email to kubevirt-dev@googlegroups.com, a message in the Kubernetes Slack channel \#kubevirt-dev or add a topic to the agenda for the next community meeting. Now we dont have HW for CI pipelines which may prevent your patch from getting merged, you could also ask the community if they can do an exception. 

For upgrading fedora-with-test-tooling, what about opening another PR to handle it

ajcaldelas commented:
Does this still not work with the pflash loader?

zhencliu commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -732,6 +732,10 @@ type SEVPolicy struct {
+	// SEV-SNP is required.
+	// Defaults to false.
+	// +optional
+	SecureNestedPaging *bool `json:"secureNestedPaging,omitempty"`
```
hi @ymjing , I happened to find that SNP has different attributes from SEV/SEV-ES, refer to docs/formatdomain.rst of libvirt source code:
     <launchSecurity type='sev-snp' authorKey='yes' vcek='no'>
       <cbitpos>47</cbitpos>
       <reducedPhysBits>1</reducedPhysBits>
       <policy>0x00030000</policy>
       <guestVisibleWorkarounds>...</guestVisibleWorkarounds>
       <idBlock>...</idBlock>
       <idAuth>...</idAuth>
       <hostData>...</hostData>
     </launchSecurity>

Besides the common attributes, e.g. cbitpos, shared with SEV/SEV-ES, it also has specialized attributes, e.g. hostData, and SNP's policy is different too, so do you think it's reasonable to introduce a SNP structure instead of taking SNP as a SEVPolicy?

ymjing commented:
Hi I've found a workaround so I don't need to use Kubevirt in my project. 

Hopefully this PR can still be useful for others. However, I don't have the bandwidth to work on this PR for now. And appreicate it if someone else can take over and finish any remaining work.

This PR can be closed if there's no interest in the next 30 days.

ajcaldelas commented:
I'd be happy to take over this PR, what do you need me to do take over?

ymjing commented:
> I'd be happy to take over this PR, what do you need me to do take over?

Thanks! If you are a member of the `kubevirt` org you probbably can directly push to this PR. Otherwise feel free to fork it.

zhencliu commented:
> Hi I've found a workaround so I don't need to use Kubevirt in my project.
> 
> Hopefully this PR can still be useful for others. However, I don't have the bandwidth to work on this PR for now. And appreicate it if someone else can take over and finish any remaining work.
> 
> This PR can be closed if there's no interest in the next 30 days.

Hi, this PR is indeed useful for us and thank you for this valuable contribution!

zhencliu commented:
> I'd be happy to take over this PR, what do you need me to do take over?

Hi Alan, thank you for taking over this, may I ask your plan? I would like to have a try at your new PR then.

alancaldelas commented:
@ymjing @zhencliu I just wanted to confirm that you guys had this working, I ran into an issue where the labels were not being set to "true" but just "". I changed how the label is set to "true" instead of "" and that did work for me. 
I didn't run this in a Kind cluster, I built the manifests and applied to a running cluster.

zhencliu commented:
> @ymjing @zhencliu I just wanted to confirm that you guys had this working, I ran into an issue where the labels were not being set to "true" but just "". I changed how the label is set to "true" instead of "" and that did work for me. I didn't run this in a Kind cluster, I built the manifests and applied to a running cluster.

hi @alancaldelas, yes, I did it in the same way as what you did, label the node with snp=true, then the node can be scheduled. The issue is when I destroy the snp vm, and start it again, I have to re-label the node with snp=true again, I am not sure if this is a bug
